src/components/app/MenuButtons/Publish.js

```diff
@@ -378,30 +350,76 @@ const profileSharingOptions: ExplicitConnectOptions<
   },
   component: MenuButtonsPublishImpl,
 };
-const MenuButtonsPublishConnected = explicitConnect(profileSharingOptions);
+export const MenuButtonsPublish = explicitConnect(options);
+
+type DownloadSizeProps = {|
+  // The generation is a number that only increases, and does so each time a profile
+  // sanitization option changes, or a new profile is generated. It should be increased
+  // any time a new sanitized profile is generated. This is how we know to update
+  // this component.
+  +generation: number,
+  +downloadSizePromise: Promise<string>,
+|};
 
-type DownloadSizeProps = {| +downloadSizePromise: Promise<string> |};
+type DownloadSizeState = {|
+  prevGeneration: number | null,
+  downloadSize: string | null,
+|};
 
 /**
- * This class should be correctly keyed so that it never updates.
+ * The DownloadSize handles unpacking the downloadSizePromise. It does
+ * so by deriving the state from the props, and utilizing a generation value.
+ * See the generation props for a more detailed explanation.
  */
 class DownloadSize extends React.PureComponent<
   DownloadSizeProps,
-  {| downloadSize: string | null |}
+  DownloadSizeState
 > {
   _isMounted: boolean = true;
-  state = { downloadSize: null };
 
-  componentDidMount() {
+  state = {
+    prevGeneration: null,
+    downloadSize: null,
+  };
+
+  /**
+   * The props changed, derive the new state from them. This function will invalidate
+   * the previous compressedProfileObjectUrl if the generation is different.
+   */
+  static getDerivedStateFromProps(
+    props: DownloadSizeProps,
+    state: DownloadSizeState
+  ): null | DownloadSizeState {
+    if (props.generation !== state.prevGeneration) {
+      // The generation value changed, invalidate the download size.
+      return {
+        prevGeneration: props.generation,
+        downloadSize: null,
```

Comment:
by setting `downloadSize` to `null`, it disappears, and I think this "flickers" a bit. Instead if we don't set it to `null`, it will change only after the promise is resolved. I think the result is cleaner, but admittedly this is subjective. What do you think?